src/Raven.Server/ServerWide/Commands/Indexes/PutIndexCommand.cs

```diff
@@ -50,6 +58,31 @@ public override string UpdateDatabaseRecord(DatabaseRecord record, long etag)
             return null;
         }
 
+        private bool InitializeStaggeredDeployment(DatabaseRecord record)
+        {
+            // What happens here though? No members.. do we resend the command? How do we make sure this will run again?
+            if (record.Topology.Members.Count == 0)
+                return false;
+
+            record.StaggeredIndexes ??= new Dictionary<string, Dictionary<string, StaggeredIndexStatus>>();
+
+            if (record.StaggeredIndexes.TryAdd(Definition.Name, new Dictionary<string, StaggeredIndexStatus>()) == false)
+            {
+                //staggered index already being deployed, no need to add it to indexes again
+                return false;
+            }
+
+            // It's guaranteed that topology members are ordered the same on all nodes
+            var lastMember = record.Topology.Members.Last();
+
+            foreach (var member in record.Topology.Members)
```

Comment:
What about `Rehab` and `Promotable` ? I think we need to populate them with `Running` ? or is it taken care of in `ShouldSkipNode`?

Reply:
One thing to consider about `Rehab`/`Promotable` it might be that their mentor doesn't start to run this index, so we promote this node to member. 
Not sure if we want to say that if we have a staggered index, we want to wait for its completion before promoting.

Reply:
But Rehab / Promotable aren't "live", shouldn't they start running all the indexes ASAP?
And yes, we need to be careful about not moving the mentor to rehab because of that.

Reply:
My initial thinking was if the index was defined to be staggered, it's probably a heavy index. 
Why are nodes in rehab? Probably because they had performance or connection trouble. Should we take these "weak" nodes and throw the index at them? Don't we want to let them recover first, become members, and only then let them start the indexing?
It will be much easier to have them run the index immediately but doesn't that defeat the purpose?

Reply:
The problem is that they now go to full members, but don't have this index deployed, that will surprise the users,likely.

Reply:
I still believe that running rehabs immediately defeats the purpose of deploying in a staggered manner and contradicts the "guarantee" of this feature to have each node run separately to maximize the cluster's high availability.

I suggest that if we have rehabs/promotables we can also add them to the staggered dictionary in a pending state but they will be executed last after all the members are done. So in each stage, the finished node chooses a member and if none are left, a rehab node will be chosen. In this way, we don't promote to be a full member until the index is done so no surprise there.